R/step_climate.R

```diff
@@ -335,18 +345,75 @@ print.step_climate <- function(x, width = max(20, options()$width - 30), ...) {
 #' @param window_size the number of .idx entries before and after to include in
 #'   the aggregation
 #' @param modulus the maximum value of `.idx`
+#' @importFrom lubridate %m-%
 roll_modular_multivec <- function(col, .idx, weights, aggr, window_size, modulus) {
   tib <- tibble(col = col, weights = weights, .idx = .idx) |>
     arrange(.idx) |>
     tidyr::nest(data = c(col, weights), .by = .idx)
-  out <- double(nrow(tib))
+  out <- double(modulus + 1)
   for (iter in seq_along(out)) {
+    # +1 from 1-indexing
     entries <- (iter - window_size):(iter + window_size) %% modulus
     entries[entries == 0] <- modulus
+    # note that because we are 1-indexing, we're looking for indices that are 1
+    # larger than the actual day/week in the year
+    if (modulus == 365) {
+      # we need to grab just the window around the leap day on the leap day
+      if (iter == 366) {
+        # there's an extra data point in front of the leap day
+        entries <- (59 - window_size):(59 + window_size - 1) %% modulus
+        entries[entries == 0] <- modulus
+        # adding in the leap day itself
+        entries <- c(entries, 999)
+      } else if ((59 %in% entries) || (60 %in% entries)) {
+        # if we're on the Feb/March boundary for daily data, we need to add in the
+        # leap day data
+        entries <- c(entries, 999)
+      }
+    } else if (modulus == 52) {
+      # we need to grab just the window around the leap week on the leap week
+      if (iter == 53) {
+        entries <- (53 - window_size):(53 + window_size - 1) %% 52
+        entries[entries == 0] <- 52
+        entries <- c(entries, 999)
+      } else if ((52 %in% entries) || (1 %in% entries)) {
+        # if we're on the year boundary for weekly data, we need to add in the
+        # leap week data (which is the extra week at the end)
+        entries <- c(entries, 999)
+      }
+    }
     out[iter] <- with(
-      purrr::list_rbind(tib$data[entries]),
+      purrr::list_rbind(tib |> filter(.idx %in% entries) |> pull(data)),
```

Comment:
I prefer the "base R" pipe `|>` as well, but unless we depend on R (>=4.1), we can't use it.
```suggestion
      purrr::list_rbind(tib %>% filter(.idx %in% entries) %>% pull(data)),
```